wasm/jit/engine.go

```diff
@@ -0,0 +1,329 @@
+	} else if index, ok := e.compiledWasmFunctionIndex[f]; ok {
+		f := e.compiledWasmFunctions[index]
+		e.exec(f)
+	} else {
+		err = fmt.Errorf("invalid function")
+		return
+	}
+	returns = make([]uint64, len(f.Signature.ReturnTypes))
+	for i := range returns {
+		returns[len(returns)-1-i] = e.pop()
+	}
+	return
+}
+
+// PreCompile implements wasm.Engine for engine.
+// Here we assign unique ids to all the function instances,
+// so we can reference it when we compile each function instance.
+func (e *engine) PreCompile(fs []*wasm.FunctionInstance) error {
+	var newUniqueHostFunctions, newUniqueWasmFunctions int
+	for _, f := range fs {
+		if f.HostFunction != nil {
+			if _, ok := e.hostFunctionIndex[f]; ok {
+				continue
+			}
+			id := int64(len(e.hostFunctionIndex))
+			e.hostFunctionIndex[f] = id
+			newUniqueHostFunctions++
+		} else {
+			if _, ok := e.compiledWasmFunctionIndex[f]; ok {
+				continue
+			}
+			id := int64(len(e.compiledWasmFunctionIndex))
+			e.compiledWasmFunctionIndex[f] = id
+			newUniqueWasmFunctions++
+		}
+	}
+	e.hostFunctions = append(
+		e.hostFunctions,
+		make([]func(ctx *wasm.HostFunctionCallContext), newUniqueHostFunctions)...,
+	)
+	e.compiledWasmFunctions = append(
+		e.compiledWasmFunctions,
+		make([]*compiledWasmFunction, newUniqueWasmFunctions)...,
+	)
+	return nil
+}
+
+func (e *engine) Compile(f *wasm.FunctionInstance) error {
+	if f.HostFunction != nil {
+		id := e.hostFunctionIndex[f]
+		if e.hostFunctions[id] != nil {
+			// Already compiled.
+			return nil
+		}
+		hf := func(ctx *wasm.HostFunctionCallContext) {
+			tp := f.HostFunction.Type()
+			in := make([]reflect.Value, tp.NumIn())
+			for i := len(in) - 1; i >= 1; i-- {
+				val := reflect.New(tp.In(i)).Elem()
+				raw := e.pop()
+				kind := tp.In(i).Kind()
+				switch kind {
+				case reflect.Float64, reflect.Float32:
+					val.SetFloat(math.Float64frombits(raw))
+				case reflect.Uint32, reflect.Uint64:
+					val.SetUint(raw)
+				case reflect.Int32, reflect.Int64:
+					val.SetInt(int64(raw))
+				}
+				in[i] = val
+			}
+			val := reflect.New(tp.In(0)).Elem()
+			val.Set(reflect.ValueOf(ctx))
+			in[0] = val
+			for _, ret := range f.HostFunction.Call(in) {
+				switch ret.Kind() {
+				case reflect.Float64, reflect.Float32:
+					e.push(math.Float64bits(ret.Float()))
+				case reflect.Uint32, reflect.Uint64:
+					e.push(ret.Uint())
+				case reflect.Int32, reflect.Int64:
+					e.push(uint64(ret.Int()))
+				default:
+					panic("invalid return type")
+				}
+			}
+		}
+		e.hostFunctions[id] = hf
+	} else {
+		id := e.compiledWasmFunctionIndex[f]
+		if e.compiledWasmFunctions[id] != nil {
+			// Already compiled.
+			return nil
+		}
+		cf, err := e.compileWasmFunction(f)
+		if err != nil {
+			return fmt.Errorf("failed to compile Wasm function: %w", err)
+		}
+		e.compiledWasmFunctions[id] = cf
+	}
+	return nil
+}
+
+func NewEngine() wasm.Engine {
+	return newEngine()
+}
+
+const initialStackSize = 100
```

Comment:
is there a max?

Reply:
no max for now. Later we add when we try to pass the spec tests!